### PR TITLE
317 update devcontainer for reproducible build

### LINKED ISSUE #318: pmtiles_buildings target happening after save_outputs
As shown below. Heads-up @mem48 

image

### PR DESCRIPTION
**Pull request by atumscott**
- Update with more dependencies (#317)
- Update install deps sh file
- add requirements.txt and modify Dockerfile
- Remove unnused files
- Separate install from loading pkgs
- Change target checks, close #318
- Update with working target order (#318)
- Install tippecanoe
- add changes (#321)
- Update odjitter location
- save changes
- Update install script

### CODE REVIEW

**Comment by Robinlovelace** [comment]
As discussed with @wangzhao0217 the code fails because `odjitter` could not be found. I have an idea what may be causing this: `~/.cargo` is added to the path incorrectly.

**Comment by Robinlovelace** [comment]
See commit above and https://unix.stackexchange.com/questions/26047/how-to-correctly-add-a-path-to-path for more info

**Review by Robinlovelace — APPROVED** [review]

**Comment by Robinlovelace** [comment]
Heads-up @wangzhao0217 I think your solution was close. Seems to work with this:

```r
old_path <- Sys.getenv("PATH")
Sys.setenv(PATH = paste(old_path, "/root/.cargo/bin", sep = ":"))
```

**Comment by Robinlovelace** [comment]
@wangzhao0217 can you try pulling down the latest version of the code and trying to run the code again?

**Comment by Robinlovelace** [comment]
One more point @wangzhao0217 you should ensure you have the Renviron:

```r
file.copy("inputdata/.Renviron", ".")
```

**Comment by wangzhao0217** [comment]
Morning, I am working on it, but not working so far, seems like the rebuild dev container not working somehow

**Comment by Robinlovelace** [comment]
@wangzhao0217 what issues are you seeing?

**Comment by wangzhao0217** [comment]
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 2: $'\r': command not found
cat: '/etc/os-release'$'\r': No such file or directory
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 5: $'\r': command not found
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 8: $'\r': command not found
': No such file or directorye 'code/install.R
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 11: $'\r': command not found
E: Invalid operation update
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package software-properties-common
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package python3-pip
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 18: $'\r': command not found
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 20: pip: command not found
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 21: $'\r': command not found
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 23: $'\r': command not found
info: downloading installer
' which wasn't expected, or isn't valid in this context

` as a value rather than a flag, use `-- -

USAGE:
    rustup-init [OPTIONS]

For more information try --help

/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 26: $'\r': command not found
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 29: $'\r': command not found
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 32: $'\r': command not found
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 34: cargo: command not found
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 35: $'\r': command not found
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 37: cd: $'/tmp\r': No such file or directory
fatal: destination path 'tippecanoe' already exists and is not an empty directory.
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 39: cd: $'tippecanoe\r': No such file or directory
make: the '-j' option requires a positive integer argument
Usage: make [options] [target] ...
Options:
  -b, -m                      Ignored for compatibility.
  -B, --always-make           Unconditionally make all targets.
  -C DIRECTORY, --directory=DIRECTORY
                              Change to DIRECTORY before doing anything.
  -d                          Print lots of debugging information.
  --debug[=FLAGS]             Print various types of debugging information.
  -e, --environment-overrides
                              Environment variables override makefiles.
  -E STRING, --eval=STRING    Evaluate STRING as a makefile statement.
  -f FILE, --file=FILE, --makefile=FILE
                              Read FILE as a makefile.
  -h, --help                  Print this message and exit.
  -i, --ignore-errors         Ignore errors from recipes.
  -I DIRECTORY, --include-dir=DIRECTORY
                              Search DIRECTORY for included makefiles.
  -j [N], --jobs[=N]          Allow N jobs at once; infinite jobs with no arg.
  -k, --keep-going            Keep going when some targets can't be made.
  -l [N], --load-average[=N], --max-load[=N]
                              Don't start multiple jobs unless load is below N.
  -L, --check-symlink-times   Use the latest mtime between symlinks and target.
  -n, --just-print, --dry-run, --recon
                              Don't actually run any recipe; just print them.
  -o FILE, --old-file=FILE, --assume-old=FILE
                              Consider FILE to be very old and don't remake it.
  -O[TYPE], --output-sync[=TYPE]
                              Synchronize output of parallel jobs by TYPE.
  -p, --print-data-base       Print make's internal database.
  -q, --question              Run no recipe; exit status says if up to date.
  -r, --no-builtin-rules      Disable the built-in implicit rules.
  -R, --no-builtin-variables  Disable the built-in variable settings.
  -s, --silent, --quiet       Don't echo recipes.
  --no-silent                 Echo recipes (disable --silent mode).
  -S, --no-keep-going, --stop
                              Turns off -k.
  -t, --touch                 Touch targets instead of remaking them.
  --trace                     Print tracing information.
  -v, --version               Print the version number of make and exit.
  -w, --print-directory       Print the current directory.
  --no-print-directory        Turn off -w, even if it was turned on implicitly.
  -W FILE, --what-if=FILE, --new-file=FILE, --assume-new=FILE
                              Consider FILE to be infinitely new.
  --warn-undefined-variables  Warn when an undefined variable is referenced.

This program built for x86_64-pc-linux-gnu
Report bugs to <bug-make@gnu.org>
'.  Stop. No rule to make target 'install
/workspaces/npt/.devcontainer/install-additional-dependencies.sh: line 42: tippecanoe: command not found

**Comment by Robinlovelace** [comment]
@wangzhao0217 did you get it working?

File: _targets.R [hunk 1]
```diff
@@ -145,13 +139,20 @@ list(
...
+    }
     odj = odjitter::jitter(
       od = od,
       zones = z,
       subpoints_origins = subpoints_origins,
       subpoints_destinations = subpoints_destinations,
       disaggregation_threshold = 30,
-      deduplicate_pairs = FALSE
+      deduplicate_pairs = FALSE,
+      odjitter_location = "/root/.cargo/bin/odjitter"
```
Comment (Robinlovelace) [inline _targets.R:155]: To remove.

**Review by Robinlovelace — APPROVED** [review]

**Comment by Robinlovelace** [comment]
Ready for you to merge after any additional minor changes pushed @wangzhao0217

**Review by Robinlovelace — APPROVED** [review]

File: _targets.R (hunk 1 above)
Comment (wangzhao0217) [reply _targets.R]: quick one: These lines need to stay, removing it will cause the odjitter to miss again.

File: _targets.R [hunk 2]
```diff
@@ -1300,53 +1288,75 @@ tar_target(pmtiles_rnet, {
...
       filter(rowSums(is.na(select(., all_of(columns_to_check)))) != length(columns_to_check))
-        
-    # Perform a spatial join to find geometries within 'rnet_x_buffer'
-    within_join <- st_join(rnet_yp, rnet_x_buffer, join = st_within)
+
+    # Selecting only the geometry column from the 'rnet_merged_all' dataset.
+    rnet_merged_all_only_geometry = rnet_merged_all %>% select(geometry)
+
+    # Merging all geometries into a single geometry using st_union from the sf package.
+    rnet_merged_all_union = sf::st_union(rnet_merged_all_only_geometry)
```
Comment (Robinlovelace) [inline _targets.R:1298]: Looks good. Did it work?

**Review by Robinlovelace — COMMENTED** [review]
I'd like to test this before merging but so far: looks good.

**Comment by wangzhao0217** [comment]
yes, please test it, and I am testing now

File: code/install.R [hunk 3]
```diff
@@ -4,7 +4,7 @@ pkgs = packages = c(
   "stringr","sf","tidyr","data.table", "targets",
   "glue","zip","jsonlite","remotes","gert","collapse","pct",
   "readr", "future", "future.callr", "future.batchtools",
-  "bs4Dash", "DT", "gt", "pingr", "shinybusy", "shinyWidgets"
+  "bs4Dash", "DT", "gt", "pingr", "shinybusy", "shinyWidgets","geos"
```
[the remark this answers is not in the document]
Comment (Robinlovelace) [reply code/install.R:7]: OK great, already done!

**Comment by wangzhao0217** [comment]
already removed

**Comment by wangzhao0217** [comment]
Hi Joey, you probably missing this
file, /workspaces/npt/inputdata/SIMD/simd2020_withgeog.zip
[image: image.png]

**Comment by Robinlovelace** [comment]
I can help debug this...

**Comment by Robinlovelace** [comment]
Any luck now @joeytalbot ? Fingers crossed!

**Comment by Robinlovelace** [comment]
Thanks for testing Joey. @wangzhao0217 can you reproduce this? Wonder if it's due to the small number of routes (200 in parameters.json).

**Comment by Robinlovelace** [comment]
Maybe try `stplanr::rnet_merge()`?

**Comment by Robinlovelace** [comment]
> > Maybe try `stplanr::rnet_merge()`?
> 
> Good idea but this returns us to the original error

This one?

```
Error:
! Error running targets::tar_make()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: Assigned data `value` must be compatible with existing data.
✖ Existing data has 474 rows.
✖ Assigned data has 0 rows.
ℹ Only vectors of size 1 are recycled.
Caused by error in `vectbl_recycle_rhs_rows()`:
! Can't recycle input of size 0 to size 474.
```

**Comment by Robinlovelace** [comment]
Can you reproduce the error outside of targets? You should be able to do that with `tar_load(object)` for every object required for that target.

**Comment by Robinlovelace** [comment]
> > Can you reproduce the error outside of targets? You should be able to do that with `tar_load(object)` for every object required for that target.
> 
> I've done this and it reached the same error, as I described above

Good progress. So it's the `rnet_merge()` line that's failing? I suggest trying to run the build but with more routes e.g. 1000 and maybe changing other settings in parameters.json.

**Comment by wangzhao0217** [comment]
@joeytalbot could you pls try tar_make(simplify_network) and let me know the error?

**Comment by Robinlovelace** [comment]
Just had an idea: maybe we needed to install `stplanr`: always worth updating packages when something doesn't work.

**Comment by Robinlovelace** [comment]
I just ran

```r
remotes::install_dev("stplanr")
```

**Comment by Robinlovelace** [comment]
Good news: that fixed the issue that @joeytalbot identified!

Bad news: there's another error message, unrelated to network simplification:

```
✖ error target zones_tile
▶ end pipeline [10.161 seconds]
Warning messages:
1: st_centroid assumes attributes are constant over geometries 
2: 1 targets produced warnings. Run targets::tar_meta(fields = warnings, complete_only = TRUE) for the messages. 
Error:
! Error running targets::tar_make()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: `x` and `y` must share the same src.
ℹ `x` is a <tbl_df/tbl/data.frame> object.
ℹ `y` is `NULL`.
ℹ Set `copy = TRUE` if `y` can be copied to the same source as `x` (may be
  slow).
```

**Comment by Robinlovelace** [comment]
Finally reproduced this locally so going to merge. Heads-up @mem48 you'll need to move the `inputdata` folder into the `npt` after this for things to work.

**Pull request merged**